File: src/ru/net/ndt/locky37/main/Main.java
package ru.net.ndt.locky37.main;

import java.io.IOException;
import java.nio.file.*;
import java.nio.file.attribute.BasicFileAttributes;

import ru.net.ndt.locky37.zip.Compress;
import ru.net.ndt.locky37.socket.*;

/**
 * This program copies a whole directory (including its sub files and
 * sub directories) to another, using the Java NIO API.
 *
 * @author www.codejava.net
 */

class Main extends SimpleFileVisitor<Path> {
    private Path sourceDir;
    private Path targetDir;

    private Main(Path sourceDir, Path targetDir) {
        this.sourceDir = sourceDir;
        this.targetDir = targetDir;
    }

    @Override
    public FileVisitResult visitFile(Path file, BasicFileAttributes attributes) {

        try {
            Path targetFile = targetDir.resolve(sourceDir.relativize(file));
            Files.copy(file, targetFile);
        } catch (IOException ex) {
            System.err.println("IO Error File");
        }

        return FileVisitResult.CONTINUE;
    }

    @Override
    public FileVisitResult preVisitDirectory(Path dir, BasicFileAttributes attributes) {
        try {
            Path newDir = targetDir.resolve(sourceDir.relativize(dir));
            Files.createDirectory(newDir);
        } catch (IOException ex) {
            System.err.println("Specify a directory");
            System.exit(0);
        }

        return FileVisitResult.CONTINUE;
    }

    public static void main(String[] args) throws IOException {

        Path sourceDir;
        Path targetDir;

        if (args.length == 0) {
            System.out.println("Example: java -jar backup_service.jar SRC_DIR DST_DIR");
            System.out.println("Example with ZIP: java -jar backup_service.jar -c SRC_DIR DST_ZIP_FILE");
            System.out.println("Example with ZIP: java -jar backup_service.jar -server DST_DIR");
            System.out.println("Example with ZIP: java -jar backup_service.jar -client hostname SRC_DIR");
            System.exit(0);
        } else if (args[0].equals("-server")) {
            System.out.println("modification = server");
            try {
                while (true) {
                    ServerZip server = new ServerZip();
                    server.serverZip(args[1]);
                    System.out.println("Server Copy with compression complete!");
                }
            } catch (IOException e) {
                e.printStackTrace();
            }
        } else if (args[0].equals("-client")) {
            System.out.println("modification = client");

            ClientZip client = new ClientZip();
            client.clientZip(args[1],args[2]);
            System.out.println("Client Copy with compression complete!");

        } else if (args[0].equals("-c")) {
            System.out.println("modification = c");
            sourceDir = Paths.get(args[1]);
            targetDir = Paths.get(args[2]);

            String dir = String.valueOf(sourceDir);
            String zipDirName = String.valueOf(targetDir);

            Compress zipFiles = new Compress();
            zipFiles.zipDirectory(dir, zipDirName);
            System.out.println("Copy with compression complete!");
        } else if (args[1].length() != 0 && args[2].length() != 0) {
            sourceDir = Paths.get(args[0]);
            targetDir = Paths.get(args[1]);
            Files.walkFileTree(sourceDir, new Main(sourceDir, targetDir));
            System.out.println("Copy complete!");
        } else System.out.println("Nothing");

        /*ServerZip server = new ServerZip();
        server.serverZip(args[0]);*/
    }
}

File: src/ru/net/ndt/locky37/main/Main.kt
package ru.net.ndt.locky37.main

import java.io.File
import java.io.FileOutputStream
import java.io.IOException
import java.nio.file.*
import java.util.zip.ZipOutputStream

import ru.net.ndt.locky37.zip.Compress
import ru.net.ndt.locky37.socket.*
import kotlin.system.exitProcess
import java.nio.file.Paths
import java.nio.file.FileVisitResult
import java.nio.file.Files
import java.nio.file.attribute.BasicFileAttributes
import java.nio.file.SimpleFileVisitor
import java.time.LocalDate

fun main(args: Array<String>) {

    val sourceDir: Path
    val targetDir: Path

    if (args.isEmpty()) {
        println("Example Copy Dir: java -jar backup_service.jar source... target")
        println("Example Copy Dir with ZIP: java -jar backup_service.jar -c source... target_file")
        println("Example Copy Dir with ZIP: java -jar backup_service.jar -server PORT target_dir")
        println("Example Copy Dir with ZIP: java -jar backup_service.jar -client PORT HOSTNAME source_dir")
        exitProcess(0)
    } else if (args[0] == "-server") {
        val port = Integer.parseInt(args[1])
        println("modification = server")
        val server = ServerZip()
        println(port)

        try {
            while (true) {
                //val server = ServerZip()
                server.serverZip(args[2], port)
                println("Server Copy with compression complete!")
            }
        } catch (e: IOException) {
            e.printStackTrace()
        }

    } else if (args[0] == "-client") {
        val port = Integer.parseInt(args[1])
        sourceDir = Paths.get(args[3])
        println("modification = client")
        println(port)

        val date = LocalDate.now()

        val temp = System.getProperty("java.io.tmpdir")
        val tempFile = File("${temp}copy\\${sourceDir.fileName}$date.zip")

        if (!tempFile.exists()) {
            makeDir(temp)

            //sourceDir = Paths.get(args[3])
            targetDir = Paths.get("${temp}copy\\${sourceDir.fileName}$date.zip")
            val fileToZip = File(sourceDir.toString())
            val fos = FileOutputStream(targetDir.toString())
            val zipOut = ZipOutputStream(fos)

            val zipFiles = Compress()
            zipFiles.zipFolder(fileToZip, fileToZip.name, zipOut)
            zipOut.close()
            fos.close()
            println("Compression complete!")
        }

        val client = ClientZip()
        try {
            client.clientZip(port, args[2], "${temp}copy\\")
            deleteDir(Paths.get("${temp}copy\\"))
            println("Client Copy with compression complete!")
        } catch (e: ArrayIndexOutOfBoundsException) {
            println("Please enter hostname")
        }

    } else if (args[0] == "-c") {
        println("modification = c")
        //String sourceFile = "zipTest";
        sourceDir = Paths.get(args[1])
        targetDir = Paths.get(args[2])
        val fileToZip = File(sourceDir.toString())
        val fos = FileOutputStream(targetDir.toString())
        val zipOut = ZipOutputStream(fos)

        val zipFiles = Compress()
        zipFiles.zipFolder(fileToZip, fileToZip.name, zipOut)
        zipOut.close()
        fos.close()
        println("Copy with compression complete!")
    } else if (args[0].isNotEmpty() && args[1].isNotEmpty()) {
        sourceDir = Paths.get(args[0])
        targetDir = Paths.get(args[1])
        Copy(sourceDir, targetDir)
        //Files.walkFileTree(sourceDir, new Main(sourceDir, targetDir));
        println("Copy complete!")
    }
}

fun deleteDir(getDir: Path) {
    Files.walkFileTree(getDir, object : SimpleFileVisitor<Path>() {
        override fun postVisitDirectory(
                dir: Path, exc: IOException?): FileVisitResult {
            Files.delete(dir)
            return FileVisitResult.CONTINUE
        }

        override fun visitFile(
                file: Path, attrs: BasicFileAttributes): FileVisitResult {
            Files.delete(file)
            return FileVisitResult.CONTINUE
        }
    })
}

fun makeDir(temp: String) {
    val mkdir = File("${temp}copy")
    if (!mkdir.exists()) {
        if (mkdir.mkdir()) {
            println("Directory is created!")
        } else {
            println("Failed to create directory!")
        }
    }
}
